svcb.go

```diff
@@ -0,0 +1,802 @@
+			}
+			if err := keyValue.parse(val); err != nil {
+				return &ParseError{"", err.Error(), l}
+			}
+			xs = append(xs, keyValue)
+		case zQuote:
+			return &ParseError{"", "SVCB key can't contain double quotes", l}
+		case zBlank:
+			canHaveNextKey = true
+		default:
+			return &ParseError{"", "bad SVCB values", l}
+		}
+		l, _ = c.Next()
+	}
+	rr.Value = xs
+	if rr.Priority == 0 && len(xs) > 0 {
+		return &ParseError{"", "SVCB aliasform can't have values", l}
+	}
+	return nil
+}
+
+// makeSVCBKeyValue returns an SVCBKeyValue struct with the key
+// or nil for reserved keys.
+func makeSVCBKeyValue(key uint16) SVCBKeyValue {
+	switch key {
+	case SVCB_MANDATORY:
+		return new(SVCBMandatory)
+	case SVCB_ALPN:
+		return new(SVCBAlpn)
+	case SVCB_NO_DEFAULT_ALPN:
+		return new(SVCBNoDefaultAlpn)
+	case SVCB_PORT:
+		return new(SVCBPort)
+	case SVCB_IPV4HINT:
+		return new(SVCBIPv4Hint)
+	case SVCB_ECHCONFIG:
+		return new(SVCBECHConfig)
+	case SVCB_IPV6HINT:
+		return new(SVCBIPv6Hint)
+	default:
+		if key == 65535 {
+			return nil
+		}
+		e := new(SVCBLocal)
+		e.KeyCode = key
+		return e
+	}
+}
+
+// SVCB RR. TODO See RFC xxxx (https://tools.ietf.org/html/draft-ietf-dnsop-svcb-https-02)
+// The one with smallest priority should be given preference.
+// Of those with equal priority, a random one should be preferred for load balancing.
+type SVCB struct {
+	Hdr      RR_Header
+	Priority uint16
+	Target   string         `dns:"domain-name"`
+	Value    []SVCBKeyValue `dns:"svcb-pairs"` // if priority == 0 this is empty
+}
+
+// HTTPS RR. Everything valid for SVCB applies to HTTPS as well
+// except that for HTTPS, HTTPS must be used
+// and HTTPS signifies that connections can be made over HTTPS.
+type HTTPS struct {
+	SVCB
+}
+
+func (rr *HTTPS) String() string {
+	return rr.SVCB.String()
+}
+
+func (rr *HTTPS) parse(c *zlexer, o string) *ParseError {
+	return rr.SVCB.parse(c, o)
+}
+
+// SVCBKeyValue defines a key=value pair for the SVCB RR type.
+// An SVCB RR can have multiple SVCBKeyValues appended to it.
+type SVCBKeyValue interface {
+	// Key returns the numerical key code.
+	Key() uint16
+	// pack returns the encoded value.
+	pack() ([]byte, error)
+	// unpack sets the value.
+	unpack([]byte) error
+	// String returns the string representation of the value.
+	String() string
+	// parse sets the value to the given string representation of the value.
+	parse(string) error
+	// copy returns a deep-copy of the pair.
+	copy() SVCBKeyValue
+	// len returns the length of value in the wire format.
+	len() uint16
+}
+
+// SVCBMandatory pair adds to required keys that must be
+// interpreted for the RR to be functional.
+// Basic use pattern for creating a mandatory option:
+//
+//	o := new(dns.SVCB)
+//	o.Hdr.Name = "."
+//	o.Hdr.Rrtype = dns.TypeSVCB
+//	e := new(dns.SVCBMandatory)
+//	e.Code = []uint16{65403}
+//	o.Value = append(o.Value, e)
+//  // Then add key-value pair for key65403
+type SVCBMandatory struct {
```

Comment:
don't really like this name, maybe wait for the RFC to be numbered and `SVBRFCXXX` or something (also don't really like that).
Is SVCBMandatory something from the I-D ?

Reply:
SVCBMandatory is the newest key-value pair in the specification.